webapp/channels/src/components/channel_bookmarks/bookmark_item.tsx

```diff
@@ -0,0 +1,242 @@
+
+import type {HTMLAttributes} from 'react';
+import React, {forwardRef, useRef} from 'react';
+import {useDispatch, useSelector} from 'react-redux';
+import {Link} from 'react-router-dom';
+import styled, {css} from 'styled-components';
+
+import type {ChannelBookmark} from '@mattermost/types/channel_bookmarks';
+import type {FileInfo} from '@mattermost/types/files';
+import type {Post} from '@mattermost/types/posts';
+
+import {getFile} from 'mattermost-redux/selectors/entities/files';
+import {getFileDownloadUrl} from 'mattermost-redux/utils/file_utils';
+
+import {openModal} from 'actions/views/modals';
+
+import ExternalLink from 'components/external_link';
+import FilePreviewModal from 'components/file_preview_modal';
+
+import {ModalIdentifiers} from 'utils/constants';
+import {getSiteURL, shouldOpenInNewTab} from 'utils/url';
+
+import type {GlobalState} from 'types/store';
+
+import BookmarkItemDotMenu from './bookmark_dot_menu';
+import BookmarkIcon from './bookmark_icon';
+
+type Props = {bookmark: ChannelBookmark};
+const BookmarkItem = <T extends HTMLAnchorElement>({bookmark}: Props) => {
+    const linkRef = useRef<T>(null);
+    const dispatch = useDispatch();
+    const fileInfo: FileInfo | undefined = useSelector((state: GlobalState) => (bookmark?.file_id && getFile(state, bookmark.file_id)) || undefined);
+
+    const open = () => {
+        linkRef.current?.click();
+    };
+
+    const handleOpenFile = (e: React.MouseEvent<HTMLAnchorElement>) => {
+        e.preventDefault();
+
+        fileInfo && dispatch(openModal({
+            modalId: ModalIdentifiers.FILE_PREVIEW_MODAL,
+            dialogType: FilePreviewModal,
+            dialogProps: {
+                post: {} as Post,
+                fileInfos: [fileInfo],
+                startIndex: 0,
+            },
+        }));
+    };
+
+    const icon = (
+        <BookmarkIcon
+            type={bookmark.type}
+            emoji={bookmark.emoji}
+            imageUrl={bookmark.image_url}
+            fileInfo={fileInfo}
+        />
+    );
+    let link;
+
+    if (bookmark.type === 'link' && bookmark.link_url) {
+        link = (
+            <DynamicLink
+                href={bookmark.link_url}
+                ref={linkRef}
+                isFile={false}
+            >
+                {icon}
+                <Label>{bookmark.display_name}</Label>
+            </DynamicLink>
+        );
+    } else if (bookmark.type === 'file' && bookmark.file_id) {
+        link = (
+            <DynamicLink
+                href={getFileDownloadUrl(bookmark.file_id)}
+                onClick={handleOpenFile}
+                ref={linkRef}
+                isFile={true}
+            >
+                {icon}
+                <Label>{bookmark.display_name}</Label>
+            </DynamicLink>
+        );
+    }
+
+    return (
+        <Chip>
+            {link}
+            <BookmarkItemDotMenu
+                bookmark={bookmark}
+                open={open}
+            />
```

Comment:
this could show a chip with a dot menu without any item as the `if/else` if statement could not be met at some point